### PR TITLE
Refactor: use ValueOrContainer for diff event

### PR DESCRIPTION
**Pull request by Leeeon233**
return `ValueOrContainer` in diff event for a better developer experience.

@zxch3n I found that `vitest` sometimes times out and cannot exit. Do you know why? The main branch also has this problem.

### CODE REVIEW

**Comment by zxch3n** [comment]
I haven't met the timeout problem for a while after I upgraded it. Can you try to reinstall the deps in loro-js?

**Comment by zxch3n** [comment]
Can you attach a benchmark comparison? I wonder whether it has too much performance penalty.

**Comment by Leeeon233** [comment]
image

File: crates/loro-internal/src/event.rs [hunk 1]
```diff
@@ -74,7 +125,7 @@ pub(crate) struct InternalContainerDiff {
 #[derive(Debug, Clone, EnumAsInner)]
 pub(crate) enum DiffVariant {
     Internal(InternalDiff),
-    External(Diff),
+    External(UnresolvedDiff),
```
Comment (zxch3n) [inline crates/loro-internal/src/event.rs:128]: Can we use resolved diff directly here?

Comment (zxch3n) [reply crates/loro-internal/src/event.rs:128]: > are difficult to obtain at that time.

Why? We can store both of them in `DocState`, can't we?

File: crates/loro-internal/src/state.rs [hunk 2]
```diff
@@ -112,18 +134,28 @@ impl State {
...
     #[inline]
-    pub fn new(arena: SharedArena) -> Self {
+    pub fn new_arc(
+        arena: SharedArena,
+        global_txn: Weak<Mutex<Option<Transaction>>>,
+    ) -> Arc<Mutex<Self>> {
         let peer = DefaultRandom.next_u64();
         // TODO: maybe we should switch to certain version in oplog?
-        Self {
+        let state = Arc::new(Mutex::new(Self {
```
Comment (zxch3n) [inline crates/loro-internal/src/state.rs:143]: You can use `Arc::new_cyclic` here

File: crates/loro-internal/src/state.rs [hunk 3]
```diff
@@ -40,6 +42,9 @@ pub struct DocState {
     pub(super) states: FxHashMap<ContainerIdx, State>,
     pub(super) arena: SharedArena,
 
+    // resolve event stuff
+    weak_state: Weak<Mutex<DocState>>,
+    global_txn: Weak<Mutex<Option<Transaction>>>,
```
Comment (zxch3n) [inline crates/loro-internal/src/state.rs:47]: Maybe we can make `DocState` as the owner of `global_txn`, and remove the `global_txn` in `Loro`.

**Review by zxch3n — APPROVED** [review]

**Pull request merged**